pkg/agent/openflow/interfaces.go

```diff
@@ -0,0 +1,131 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openflow
+
+import (
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+type pipelineProtocol int
+
+const (
+	shared pipelineProtocol = iota
+	pipelineARP
+	pipelineIP
+	pipelineLast pipelineProtocol = pipelineIP
+)
+
+// A featureID is used to represent a feature which has a set of functions. Current features include:
+// - PodConnectivity: functions about connectivity for Pods in Antrea.
+// - NetworkPolicy: implementation of NetworkPolicy and functions of AntreaNetworkPolicy in Antrea.
+// - Service: implementation of Service in Antrea.
+// - Egress: implementation of Egress in Antrea.
+// - Traceflow: functions of Traceflow in Antrea.
+type featureID int
+
+const (
+	Shared featureID = iota
+	PodConnectivity
+	NetworkPolicy
+	Service
+	Egress
+	Traceflow
+)
+
+// featureTemplate is used to declare which FeatureTables will be used for this feature. Note that, if a feature wants
+// to declare a FeatureTable, the FeatureTable should be defined first.
+type featureTemplate struct {
+	stageTables map[binding.StageID][]*FeatureTable
+	feature     featureID
+}
+
+// Flexible pipeline for IPv4/IPv6 has 11 fixed stages. These stages are:
+// - ClassifierStage
+// - ValidationStage
+// - ConntrackStateStage
+// - PreRoutingStage
+// - EgressSecurityStage
+// - RoutingStage
+// - PostRoutingStage
+// - SwitchingStage
+// - IngressSecurityStage
+// - ConntrackStage
+// - OutputStage
+// Flexible pipeline for ARP has 2 fixed stages. These stages are:
+// - ValidationStage
+// - OutputStage
+// The order of stages are fixed. A stage can be empty or have one or more flow tables. To insert a flow table to flexible
+// pipeline, its corresponding FeatureTable defined first and the FeatureTable should be also declared at least by one
+// feature.
+
+// FeatureTable is the basic unit to build a flexible pipeline. A FeatureTable should only belong to a single stage and a
+// single pipeline. If a FeatureTable is declared to used by one or more features, then the FeatureTable will be used to
+// build a pipeline and its member struct `ofTable` will be initialized. Otherwise, it will be excluded from
+// building an OVS pipeline. Note that, the order of all stages is fixed, but the order of flow tables within a stage is
+// determined by priority of their corresponding FeatureTable. FeatureTable with a higher priority is assigned with a smaller
+// tableID for its corresponding flow table, which means a packet should enter the flow table before others with lower
+// priorities in the same stage.
+type FeatureTable struct {
```

Comment:
why not entirely sure why `featureTemplate` and `FeatureTable` are defined in this file (`interfaces.go`). I am also not sure that we *need* an `interface.go` given that the only interface (`feature`) is not even exported.

Reply:
I have moved the content of `interfaces.go` to `feature_templates.go`.